Histogram breaks

bbeat2782 opened the pull request:
## Describe your changes

Implemented `--breaks/-B` to ggplot histogram.

[breaks demonstration.zip](https://github.com/ploomber/jupysql/files/12222210/breaks.demonstration.zip) This notebook contains breaks demonstration and errors I implemented.

## Issue number

Closes #719 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--767.org.readthedocs.build/en/767/

<!-- readthedocs-preview jupysql end -->

bbeat2782 commented:
[breaks demonstration.zip](https://github.com/ploomber/jupysql/files/12196760/breaks.demonstration.zip)

This is the implemented behavior of breaks. @edublancas Could you take a look at the demonstration and check whether it behaves as we want when you have time?

Since I have two other issues, this does not need an immediate check.

edublancas commented on src/sql/magic_plot.py:
```diff
@@ -72,6 +79,15 @@ def execute(self, line="", cell="", local_ns=None):
         else:
             column = cmd.args.column
 
+        bin_specified = False
+        pattern = r"histogram.*( --bins | -b )"
+        if re.search(pattern, line):
+            bin_specified = True
+            if re.search(r"( --breaks | -B )", line):
+                raise exceptions.UsageError(
+                    "Both bins and breaks are specified. Must specify only one of them."
+                )
+
```
I think you can use `cmd` to check if both arguments exist.

e.g.,

```python
if cmd.args.bins and cmd.args.breaks:
    raise UsageError(...)
```

let me know if that works, otherwise we should fix it since validating arguments with regular expressions is going to be a pain if we have more complex combinations

bbeat2782 replied:
Because of the default value of bins (which is 50 for the magic command), `cmd.args.bins` always return an integer value. Thus, I used the following code instead.

```
bin_specified = " --bins " in line or " -b " in line
breaks_specified = " --breaks " in line or " -B " in line
if bin_specified and breaks_specified:
     raise exceptions.UsageError(...)
```

edublancas reviewed (changes requested):
looks great! don't forget to add some reference images and test cases

bbeat2782 commented on doc/api/magic-plot.md:
````diff
@@ -151,6 +153,12 @@ When plotting a histogram, it divides a range with the number of bins - 1 to cal
 ```
 
+### Specifying breaks
+
````
Would breaks that cause an error be the ill-defined cases?

edublancas replied:
yes, any conditions that leads you to raise an error is what I meant by ill-defined

bbeat2782 replied:
Some context added in magic-plot.md

yafimvo replied:
I agree. I think we can elaborate a bit more. In this [example](https://www.tutorialspoint.com/how-to-define-the-breaks-for-a-histogram-using-ggplot2-in-r), they show a nice example of what can be done with breaks.

also, I think we should add a note that we can't use both `--bins` and `--breaks` since it looks like ggplot2 allows it (why can't we use both of them?)

bbeat2782 replied:
Added extra information about what breaks are used for and that we can't use both `--bins` and `--breaks`.

tonykploomber reviewed (approved):
LG, very comprehensive test case covered, GJ

neelasha23 commented on src/sql/plot.py:
```diff
@@ -392,6 +394,21 @@ def histogram(
+    if isinstance(breaks, list):
+        if len(breaks) < 2:
+            raise ValueError(
+                f"Breaks given : {breaks}. When using breaks, please ensure "
```
`%sqlplot histogram --table penguins.csv --column body_mass_g --breaks 9000` : this prints entire stack trace : 
```
File ~/miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/plot.py:399, in histogram(payload, table, column, bins, with_, conn, category, cmap, color, edgecolor, ax, facet, breaks)
    397 if isinstance(breaks, list):
    398     if len(breaks) < 2:
--> 399         raise ValueError(
    400             f"Breaks given : {breaks}. When using breaks, please ensure "
    401             "to specify at least two points."
    402         )
    403     if not all([b2 > b1 for b1, b2 in zip(breaks[:-1], breaks[1:])]):
    404         raise ValueError(
    405             f"Breaks given : {breaks}. When using breaks, please ensure that "
    406             "breaks are strictly increasing."
    407         )

ValueError: Breaks given : [9000.0]. When using breaks, please ensure to specify at least two points.
If you need help solving this issue, send us a message: https://ploomber.io/community
```

Is this expected behaviour?

bbeat2782 replied:
Changed to `exceptions.ValueError`

neelasha23 commented:
Would it make sense to allow users to pass a range of values?

```
points = list(range(1000, 6000, 500))
%sqlplot histogram --table penguins.csv --column body_mass_g --breaks points
```

@edublancas @bbeat2782

neelasha23 commented on src/sql/plot.py:
```diff
@@ -392,6 +394,21 @@ def histogram(
+        if not all([b2 > b1 for b1, b2 in zip(breaks[:-1], breaks[1:])]):
+            raise ValueError(
+                f"Breaks given : {breaks}. When using breaks, please ensure that "
+                "breaks are strictly increasing."
```
`%sqlplot histogram --table penguins.csv --column body_mass_g --breaks 9000 8000`  Also printing entire stacktrace and raising `ValueError`. Some errors are `UsageError` and in some places it's `ValueError`.

bbeat2782 replied:
Changed `ValueError` into `exceptions.ValueError`.

In the `exceptions.py`, it says "raise UsageError when there's an issue with the user's input in a magic". But for this error, it is inside `histogram()`, which is called from `geom_histogram.py`. So this error can occur with an issue with the input in a magic, but it can also occur by calling `ggplot() + geom_histogram()` with invalid input. Thus, I used `ValueError`. If this is wrong, I will change it to `UsageError`.

bbeat2782 commented on src/sql/ggplot/geom/geom_histogram.py:
```diff
@@ -20,10 +20,11 @@ class geom_histogram(geom):
     """
 
-    def __init__(self, bins=None, fill=None, cmap=None, **kwargs):
+    def __init__(self, bins=None, fill=None, cmap=None, breaks=None, **kwargs):
```
`breaks` added to the docstring

bbeat2782 commented on src/sql/magic_plot.py:
```diff
@@ -103,12 +110,26 @@ def execute(self, line="", cell="", local_ns=None):
                 conn=None,
             )
         elif cmd.args.line[0] in {"hist", "histogram"}:
+            bin_specified = " --bins " in line or " -b " in line
```
I tried `cmd.args.bins` but because of `bins` default value, `cmd.args.bins` always contains a value (either the default or user input) regardless of user's input.

bbeat2782 commented on src/sql/plot.py:
```diff
@@ -392,6 +394,21 @@ def histogram(
+        if bins:
+            raise ValueError(
+                "Both bins and breaks are specified. Must specify only one of them."
+            )
```
Deleted the error handler in `magic_plot` and changed `ValueError` to `exceptions.ValueError`

bbeat2782 commented:
@neelasha23 

> allow users to pass a range of values

With this PR implementation, the following works.
```
points = " ".join(map(str, range(1000, 6000, 500)))
%sqlplot histogram --table penguins.csv --column body_mass_g --breaks {points}
```

> Shouldn't there be one more bin after 5000 till the highest value of body_mass_g ?

1. When specifying breaks, we are specifying a starting point, break points, and ending point. Thus, the data we query is limited to starting point <= data <= ending point. That is why the upper limit is 5000 here. This range limit is mimicking ggplot behavior in R. To avoid confusion, I added an explanation of breaks limiting range in `magic-plot.md`.

> It might help the user if we specify what the minimum point is

2. Changed the error message to `UsageError: All break points are lower than the min data point of {min point}.` (similar to the max data point)

> There would be only 1 bin if first point is less than lower datapoint and second point is higher than max value. We can provide some info to guide user when specifying points:

3. For this, do you think displaying a message will work better or giving a general info on the [docs](https://jupysql--754.org.readthedocs.build/en/754/api/magic-plot.html) will work better?

bbeat2782 commented:
@yafimvo 

> I think we should add some integration tests and make sure it works (or doesn't) on other databases (PostgreSQL, questdb, etc...)

Added test cases and images in `test_questDB.py`.

edublancas commented:
@neelasha23 @yafimvo please review again

neelasha23 commented:
> 3. For this, do you think displaying a message will work better or giving a general info on the [docs](https://jupysql--754.org.readthedocs.build/en/754/api/magic-plot.html) will work better?

Docs should be fine

neelasha23 reviewed (approved):
Looks good!

edublancas reviewed (changes requested):
looks good, one thing is missing is adding one new test here: https://github.com/ploomber/jupysql/blob/b6f07dccd356385b50f1d94768b904a0ea9f57b1/src/tests/integration/test_generic_db_operations.py#L295

you can just add an example that runs using `--breaks`, let's see if it passes. if you find compatibilities, ping me and we can decide whether we support it or not. my guess is that it'll work with the databases that we're testing there

bbeat2782 commented:
> one thing is missing is adding one new test here:

@edublancas I added `histogram-breaks`, and it passed all tests

edublancas commented:
awesome!
